sources/pg_replication/helpers.py

```diff
@@ -0,0 +1,752 @@
+                options=self.options,
+            )
+            consumer = MessageConsumer(
+                upto_lsn=self.upto_lsn,
+                pub_ops=get_pub_ops(
+                    self.options["publication_names"], self.credentials
+                ),
+                target_batch_size=self.target_batch_size,
+                include_columns=self.include_columns,
+                columns=self.columns,
+            )
+            cur.consume_stream(consumer)
+        except StopReplication:  # completed batch or reached `upto_lsn`
+            pass
+        finally:
+            cur.connection.close()
+            self.last_commit_lsn = consumer.last_commit_lsn
+            for rel_id, data_items in consumer.data_items.items():
+                table_name = consumer.last_table_schema[rel_id]["name"]
+                yield data_items[0]  # meta item with column hints only, no data
+                yield dlt.mark.with_table_name(data_items[1:], table_name)
+            self.generated_all = consumer.consumed_all
+
+
+class MessageConsumer:
+    """Consumes messages from a ReplicationCursor sequentially.
+
+    Generates data item for each `insert`, `update`, and `delete` message.
+    Processes in batches to limit memory usage.
+    Maintains message data needed by subsequent messages in internal state.
+    """
+
+    def __init__(
+        self,
+        upto_lsn: int,
+        pub_ops: Dict[str, bool],
+        target_batch_size: int = 1000,
+        include_columns: Optional[Dict[str, Sequence[str]]] = None,
+        columns: Optional[Dict[str, TTableSchemaColumns]] = None,
+    ) -> None:
+        self.upto_lsn = upto_lsn
+        self.pub_ops = pub_ops
+        self.target_batch_size = target_batch_size
+        self.include_columns = include_columns
+        self.columns = columns
+
+        self.consumed_all: bool = False
+        # data_items attribute maintains all data items
+        self.data_items: Dict[
+            int, List[Union[TDataItem, DataItemWithMeta]]
+        ] = dict()  # maps relation_id to list of data items
+        # other attributes only maintain last-seen values
+        self.last_table_schema: Dict[
+            int, TTableSchema
+        ] = dict()  # maps relation_id to table schema
+        self.last_commit_ts: pendulum.DateTime
+        self.last_commit_lsn = None
+
+    def __call__(self, msg: ReplicationMessage) -> None:
+        """Processes message received from stream."""
+        self.process_msg(msg)
+
+    def process_msg(self, msg: ReplicationMessage) -> None:
+        """Processes encoded replication message.
+
+        Identifies message type and decodes accordingly.
+        Message treatment is different for various message types.
+        Breaks out of stream with StopReplication exception when
+        - `upto_lsn` is reached
+        - `target_batch_size` is reached
+        - a table's schema has changed
+        """
+        op = (msg.payload[:1]).decode("utf-8")
```

Comment:
you could compare binary strings without decoding to make it a little bit faster. also you could put the more frequent (I/U/D) operations to the top

Reply:
Implemented in https://github.com/dlt-hub/verified-sources/pull/392/commits/7bd211b7d70de759c0809a0010e3de3fe5b66d4d.